libwallet/src/slate.rs

```diff
@@ -332,7 +331,12 @@ impl Slate {
 		K: Keychain,
 	{
 		// Whoever does this first generates the offset
-		if self.participant_data.len() == 0 {
+		// TODO: Remove HF3
+		if self.participant_data.len() == 0 && !self.is_compact() {
+			self.generate_offset(keychain, sec_key, use_test_rng)?;
+		}
+		// Always choose my part of the offset, and subtract from my excess
+		if self.is_compact() {
```

Comment:
Avoid repeating tests by
if !self.is_compact() {
  if self.participant_data.len() == 0 {
  }
} else {
}

Btw, is there no method self.participant_data.is_empty() ? comparing len() with 0 looks ugly...

Reply:
The logic is split here to allow for quick and easy deletion of 3.0.0 compatibility code once HF3 is behind us. There's a lot of this sprinkled around the code, it will be gone shortly after 4.0.0 is released.